src/t_stream.c

```diff
@@ -1338,6 +1345,132 @@ robj *createObjectFromStreamID(streamID *id) {
                         id->ms,id->seq));
 }
 
+/* Get the first or last entry ID from a non-empty stream. */
+void streamGetTipID(stream *s, streamID *id, int first) {
+    streamIterator si;
+    int64_t numfields;
+    streamID min, max;
+    min.seq = 0;
+    min.ms = 0;
+    max.seq = UINT64_MAX;
+    max.seq = UINT64_MAX;
+
+    streamIteratorStart(&si,s,&min,&max,!first);
+    streamIteratorGetID(&si,id,&numfields);
+    streamIteratorStop(&si);
+}
+
+/* Returns 1 if the ID is 0-0. */
+int streamIDEqZero(streamID *id) {
+    return !(id->ms || id->seq);
+}
+
+/* Check if a range contains an XDEL.
+ * Assumes start < end && end < s->last_id. */
+int streamIsContiguousRange(stream *s, streamID *start, streamID *end) {
+    streamID start_id, end_id;
+    streamID first_id;
+
+    if (!s->offset || streamIDEqZero(&s->xdel_max_id)) {
+        /* A newly-initialized stream or no XDELs. */
+        return 1;
+    }
+
+    /* Copy input IDs, if given, or default to min/max IDs. */
+    if (start) {
+        start_id = *start;
+    } else {
+        start_id.ms = start_id.seq = 0;
+    }
+    if (end) {
+        end_id = *end;
+    } else {
+        end_id.ms = end_id.seq = UINT64_MAX;
+    }
+
+    streamGetTipID(s,&first_id,1);
+    int cmp_first = streamCompareID(&first_id,&s->xdel_max_id);
+    if (cmp_first == 1) {
+        /* The XDEL is before the first entry. */
+        return 1;
+    }
+
+    int cmp_start = streamCompareID(&start_id,&s->xdel_max_id);
+    if (cmp_start != -1) {
+        /* The range doesn't include an XDEL. */
+        return 1;
+    }
+
+    /* The range includes an XDEL. */
+    return 0;
+}
+
+/* A helper for getting an offset for a given ID in the stream.
+ * A non-zero offset can be either logical (one before the first ID in the
+ * in the stream) or exact (the first or last IDs), but in both cases
+ * it provides a valid lag estimate. The zero offset usually means that
+ * the offset isn't available, except in the case of the 
+ * newly-initialized stream. */
+uint64_t streamGetOffset(stream *s, streamID *id) {
+    /* The offset of any ID in an empty, never-before-used stream is 0. */
+    if (!s->offset) {
+        return 0;
+    }
+
+    /* In the empty stream, if the ID is smaller or equal to the last ID,
+     * it can set to the logical start offset, which is the current offset. */
+    if (!s->length && streamCompareID(id,&s->last_id) < 1) {
+        return s->offset;
+    }
```

Comment:
not sure i got that part..

Reply:
Assume a stream with at least one entry, then delete all entries with XDEL so length ==  0. The offset of the last ID will be smaller than the next (and first in the stream) XADD, so it can be used for any ID that is LEQ the last ID.